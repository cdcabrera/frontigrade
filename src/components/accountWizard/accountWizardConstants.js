import React from 'react';
import AccountWizardStepOne from './accountWizardStepOne';
import AccountWizardStepTwo from './accountWizardStepTwo';
import AccountWizardStepThree from './accountWizardStepThree';

const awsIdentity = `{
  "Version": "2012-10-07",
  "Statement": [
    {
      "lorem": "ipsum"
    },
    {
      "dolar": "dolar"
    }
  ]
}`;

const addAccountWizardSteps = [
  {
    step: 1,
    label: '1',
    title: 'Policy',
    page: <AccountWizardStepOne identity={awsIdentity} />,
    subSteps: []
  },
  {
    step: 2,
    label: '2',
    title: 'Role',
    page: <AccountWizardStepTwo />,
    subSteps: []
  },
  {
    step: 3,
    label: '3',
    title: 'ARN',
    page: <AccountWizardStepThree />,
    subSteps: []
  }
];

const editAccountWizardSteps = [];

export { addAccountWizardSteps, editAccountWizardSteps };
