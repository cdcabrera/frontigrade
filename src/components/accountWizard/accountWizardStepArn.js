import React from 'react';
import PropTypes from 'prop-types';
import { Form, Grid } from 'patternfly-react';
import { connect, reduxTypes, store } from '../../redux';
import apiTypes from '../../constants/apiConstants';
import { FormField } from '../formField/formField';

class AccountWizardStepArn extends React.Component {
  state = {
    accountArn: '',
    accountArnError: null
  };

  onChangeAccountArn = event => {
    const { value } = event.target;
    const errorMessage = !/^\s*arn:aws:/.test(value) ? 'You must enter a valid ARN' : '';

    this.setState(
      {
        accountArn: value,
        accountArnError: errorMessage
      },
      () => this.isStepValid()
    );
  };

  // ToDo: when form valid evaluate moving to next step onEnter key-press event
  onSubmit = event => {
    event.preventDefault();
  };

  isStepValid() {
    const { accountArn, accountArnError } = this.state;
    const stepValid = accountArnError === '';
    const dispatchType = stepValid
      ? reduxTypes.account.ADD_ACCOUNT_WIZARD_STEP_ARN
      : reduxTypes.account.INVALID_ACCOUNT_WIZARD_STEP_ARN;

    const trimmedArn = stepValid ? accountArn.trim() : accountArn;

    store.dispatch({
      type: dispatchType,
      account: {
        [apiTypes.API_SUBMIT_ACCOUNT_ARN]: trimmedArn
      }
    });
  }

  render() {
    const { stepArnValid, stepArnErrorMessage } = this.props;
    const { accountArn, accountArnError } = this.state;

    let stepError = null;

    if (!stepArnValid) {
      stepError = stepArnErrorMessage;
    }

    return (
      <Form horizontal onSubmit={this.onSubmit}>
        <Form.FormGroup>
          <Grid.Col sm={12}>
            <ul className="no-margin-bottom">
              <li>
                In the <strong>IAM</strong> console, click your new role.
              </li>
              <li>
                In the <strong>Summary</strong> screen, copy the <strong>Role ARN</strong> and paste it in the following{' '}
                <strong>ARN</strong> field.
              </li>
            </ul>
          </Grid.Col>
        </Form.FormGroup>
        <FormField
          label="ARN"
          error={stepError || accountArnError}
          errorMessage={stepError || accountArnError}
          colLabel={2}
          colField={10}
        >
          <Form.FormControl type="text" name="arn" value={accountArn} onChange={this.onChangeAccountArn} />
          <Form.HelpBlock>ex: arn:aws:iam::123456789012:role/Cloud-Meter-role</Form.HelpBlock>
        </FormField>
      </Form>
    );
  }
}

AccountWizardStepArn.propTypes = {
  stepArnValid: PropTypes.bool,
  stepArnErrorMessage: PropTypes.string
};

AccountWizardStepArn.defaultProps = {
  stepArnValid: false,
  stepArnErrorMessage: null
};

const mapStateToProps = state => ({ ...state.accountWizard });

const ConnectedAccountWizardStepArn = connect(mapStateToProps)(AccountWizardStepArn);

export { ConnectedAccountWizardStepArn as default, ConnectedAccountWizardStepArn, AccountWizardStepArn };
