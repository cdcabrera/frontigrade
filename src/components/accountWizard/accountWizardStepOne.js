import React from 'react';
import PropTypes from 'prop-types';
import { Form, Grid } from 'patternfly-react';
import { connect, reduxTypes, store } from '../../redux/';
import { FormField, fieldValidation } from '../formField/formField';
import CopyField from '../copyField/copyField';

class AccountWizardStepOne extends React.Component {
  state = {
    accountName: '',
    accountNameError: null
  };

  onChangeAccountName = event => {
    // ToDo: validation for accountName may need to be completed through the API in order to avoid duplicate entries. Consider an onBlur check, along with a randomly generated name.
    const { value } = event.target;
    const errorMessage = fieldValidation.isEmpty(value) ? 'You must enter an account name' : '';

    this.setState(
      {
        accountName: value,
        accountNameError: errorMessage
      },
      () => this.isStepValid()
    );
  };

  isStepValid() {
    const { accountName, accountNameError } = this.state;
    const stepValid = accountNameError === '';

    if (stepValid) {
      store.dispatch({
        type: reduxTypes.account.ADD_ACCOUNT_WIZARD_STEPONE,
        account: {
          accountName
        }
      });
    } else {
      store.dispatch({
        type: reduxTypes.account.INVALID_ACCOUNT_WIZARD_STEPONE
      });
    }
  }

  render() {
    const { accountName, accountNameError } = this.state;
    const { policySetupConfig } = this.props;

    return (
      <React.Fragment>
        <FormField label="Account Name" error={accountNameError} errorMessage={accountNameError}>
          <Form.FormControl
            type="text"
            name="accountName"
            value={accountName}
            placeholder="Enter a name for this account"
            onChange={this.onChangeAccountName}
          />
        </FormField>
        <br />
        <br />
        <Form.FormGroup>
          <Grid.Col sm={12}>
            <ul>
              <li>
                Create a new policy in the AWS{' '}
                <a href="https://console.aws.amazon.com/iam" target="_blank" rel="noopener noreferrer">
                  Identity and Access Management
                </a>
              </li>
              <li>
                <p>In the JSON editor replace the contents with:</p>
                <CopyField multiline value={policySetupConfig} />
              </li>
              <li>
                Click <strong>Review policy</strong>.
              </li>
              <li>
                Name the policy <strong>Red-Hat-Cloud-Meter-policy</strong> and click <strong>Create policy</strong>.
              </li>
            </ul>
          </Grid.Col>
        </Form.FormGroup>
      </React.Fragment>
    );
  }
}

AccountWizardStepOne.propTypes = {
  account: PropTypes.shape({
    name: PropTypes.string
  }).isRequired,
  policySetupConfig: PropTypes.string.isRequired
};

const mapStateToProps = state => ({ ...state.accountWizard });

const ConnectedAccountWizardStepOne = connect(mapStateToProps)(AccountWizardStepOne);

export { ConnectedAccountWizardStepOne as default, ConnectedAccountWizardStepOne, AccountWizardStepOne };
