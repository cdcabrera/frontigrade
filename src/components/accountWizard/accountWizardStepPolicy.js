import React from 'react';
import PropTypes from 'prop-types';
import { Form, Grid } from 'patternfly-react';
import { awsPolicySetup } from '../../common/configuration.json';
import helpers from '../../common/helpers';
import { connect, reduxTypes, store } from '../../redux/';
import { FormField, fieldValidation } from '../formField/formField';
import CopyField from '../copyField/copyField';

class AccountWizardStepPolicy extends React.Component {
  state = {
    accountName: '',
    accountNameError: null
  };

  onChangeAccountName = event => {
    // ToDo: validation for accountName may need to be completed through the API in order to avoid duplicate entries. Consider an onBlur check, along with a randomly generated name.
    const { value } = event.target;
    const errorMessage = fieldValidation.isEmpty(value) ? 'You must enter an account name' : '';

    this.setState(
      {
        accountName: value,
        accountNameError: errorMessage
      },
      () => this.isStepValid()
    );
  };

  isStepValid() {
    const { accountName, accountNameError } = this.state;
    const stepValid = accountNameError === '';
    const dispatchType = stepValid
      ? reduxTypes.account.ADD_ACCOUNT_WIZARD_STEP_POLICY
      : reduxTypes.account.INVALID_ACCOUNT_WIZARD_STEP_POLICY;

    store.dispatch({
      type: dispatchType,
      account: {
        accountName
      }
    });
  }

  render() {
    const { accountName, accountNameError } = this.state;
    const { policySetupConfig } = this.props;

    return (
      <React.Fragment>
        <FormField label="Account Name" error={accountNameError} errorMessage={accountNameError}>
          <Form.FormControl
            type="text"
            name="accountName"
            value={accountName}
            placeholder="Enter a name for this account"
            onChange={this.onChangeAccountName}
          />
        </FormField>
        <br />
        <br />
        <Form.FormGroup>
          <Grid.Col sm={12}>
            <ul>
              <li>
                Create a new policy in the AWS{' '}
                <a href="https://console.aws.amazon.com/iam" target="_blank" rel="noopener noreferrer">
                  Identity and Access Management
                </a>
              </li>
              <li>
                <p>In the JSON editor replace the contents with:</p>
                <CopyField multiline value={policySetupConfig} />
              </li>
              <li>
                Click <strong>Review policy</strong>.
              </li>
              <li>
                Name the policy <strong>Red-Hat-Cloud-Meter-policy</strong> and click <strong>Create policy</strong>.
              </li>
            </ul>
          </Grid.Col>
        </Form.FormGroup>
      </React.Fragment>
    );
  }
}

AccountWizardStepPolicy.propTypes = {
  policySetupConfig: PropTypes.string
};

AccountWizardStepPolicy.defaultProps = {
  policySetupConfig: helpers.prettyPrintJson(awsPolicySetup)
};

const mapStateToProps = state => ({ ...state.accountWizard });

const ConnectedAccountWizardStepPolicy = connect(mapStateToProps)(AccountWizardStepPolicy);

export { ConnectedAccountWizardStepPolicy as default, ConnectedAccountWizardStepPolicy, AccountWizardStepPolicy };
