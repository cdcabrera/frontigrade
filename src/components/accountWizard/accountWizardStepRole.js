import React from 'react';
import PropTypes from 'prop-types';
import { Form, Grid, Icon } from 'patternfly-react';
import { connect, reduxActions } from '../../redux';
import apiTypes from '../../constants/apiConstants';
import helpers from '../../common/helpers';
import CopyField from '../copyField/copyField';
import Tooltip from '../tooltip/tooltip';

class AccountWizardStepRole extends React.Component {
  componentDidMount() {
    const { getSystemConfig } = this.props;

    getSystemConfig();
  }

  // ToDo: when form valid evaluate moving to next step onEnter key-press event
  onSubmit = event => {
    event.preventDefault();
  };

  render() {
    const { awsAccountId } = this.props;

    const popover = (
      <ul className="cloudmeter-popover-list">
        <li>Log in to AWS console</li>
        <li>Search Services to go to IAM</li>
        <li>Click Roles in the left nav</li>
        <li>
          Click the <strong>Create role</strong> button
        </li>
        <li>
          Click <strong>Another AWS account</strong>
        </li>
      </ul>
    );

    return (
      <Form onSubmit={this.onSubmit}>
        <Form.FormGroup>
          <Grid.Col sm={12}>
            <ul>
              <li>
                Create a new role in the AWS
                <a href="https://console.aws.amazon.com/iam" target="_blank" rel="noopener noreferrer">
                  Identity and Access Management
                </a>
                .
                <Tooltip delayShow={100} popover={popover}>
                  <Icon type="pf" name="info" size="large" />
                  <span className="sr-only">Steps when logging into AWS Identity and Access Management</span>
                </Tooltip>
              </li>
              <li>
                <p>
                  Paste this value into the Account ID field under <strong>Another AWS account</strong>:
                </p>
                <CopyField id="account-id" value={awsAccountId || ''} />
              </li>
              <li>
                In the next step, select <strong>Cloud-Meter-policy</strong>.
              </li>
              <li>
                Continue, name the role <strong>Cloud-Meter-role</strong>, and click <strong>Create role</strong>.
              </li>
            </ul>
          </Grid.Col>
        </Form.FormGroup>
      </Form>
    );
  }
}

AccountWizardStepRole.propTypes = {
  getSystemConfig: PropTypes.func,
  awsAccountId: PropTypes.string
};

AccountWizardStepRole.defaultProps = {
  getSystemConfig: helpers.noop,
  awsAccountId: null
};

const mapDispatchToProps = dispatch => ({
  getSystemConfig: () => dispatch(reduxActions.systemConfig.getSystemConfig())
});

const mapStateToProps = state => ({
  awsAccountId: state.accountWizard.configuration[apiTypes.API_SUBMIT_ACCOUNT_AWS_ID]
});

const ConnectedAccountWizardStepRole = connect(
  mapStateToProps,
  mapDispatchToProps
)(AccountWizardStepRole);

export { ConnectedAccountWizardStepRole as default, ConnectedAccountWizardStepRole, AccountWizardStepRole };
