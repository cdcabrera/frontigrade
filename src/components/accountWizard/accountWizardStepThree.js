import React from 'react';

const AccountWizardStepThree = () => <div>Lorem ipsum step three.</div>;

AccountWizardStepThree.propTypes = {};

export { AccountWizardStepThree as default, AccountWizardStepThree };
