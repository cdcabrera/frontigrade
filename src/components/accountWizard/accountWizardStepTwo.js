import React from 'react';

const AccountWizardStepTwo = () => <div>Lorem ipsum step two.</div>;

AccountWizardStepTwo.propTypes = {};

export { AccountWizardStepTwo as default, AccountWizardStepTwo };
