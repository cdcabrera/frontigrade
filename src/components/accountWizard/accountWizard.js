import React from 'react';
import PropTypes from 'prop-types';
import { Button, Icon, Wizard } from 'patternfly-react';
import { connect, reduxTypes, store } from '../../redux/';
import { addAccountWizardSteps, editAccountWizardSteps } from './accountWizardConstants';

class AccountWizard extends React.Component {
  state = {
    activeStepIndex: 0
  };

  onCancel = () => {
    const { fulfilled, error } = this.props;

    const closeWizard = () => {
      store.dispatch({
        type: reduxTypes.confirmationModal.CONFIRMATION_MODAL_HIDE
      });

      store.dispatch({
        type: reduxTypes.account.UPDATE_ACCOUNT_HIDE
      });
    };

    if (fulfilled || error) {
      closeWizard();
    } else {
      store.dispatch({
        type: reduxTypes.confirmationModal.CONFIRMATION_MODAL_SHOW,
        title: 'Cancel Add Account',
        heading: 'Are you sure you want to exit this wizard?',
        body: 'Exiting this wizard will cancel adding an account.',
        cancelButtonText: 'No',
        confirmButtonText: 'Yes',
        onConfirm: closeWizard
      });
    }
  };

  onNext = () => {};

  onBack = () => {};

  onSubmit = () => {};

  onStep = () => {
    // ToDo: wizard step map/breadcrumb/trail click, or leave disabled
  };

  renderWizardSteps() {
    const { edit } = this.props;
    const { activeStepIndex } = this.state;
    const wizardSteps = edit ? editAccountWizardSteps : addAccountWizardSteps;
    const activeStep = wizardSteps[activeStepIndex];

    return wizardSteps.map((step, stepIndex) => (
      <Wizard.Step
        key={step.title}
        stepIndex={stepIndex}
        step={step.step}
        label={step.label}
        title={step.title}
        activeStep={activeStep && activeStep.step}
      />
    ));
  }

  render() {
    const { show, edit, stepOneValid, stepTwoValid } = this.props;
    const { activeStepIndex } = this.state;
    const wizardSteps = edit ? editAccountWizardSteps : addAccountWizardSteps;

    return (
      <Wizard show={show}>
        <Wizard.Header onClose={this.onCancel} title={edit ? 'Edit Account' : 'Add Account'} />
        <Wizard.Body>
          <Wizard.Steps steps={this.renderWizardSteps()} />
          <Wizard.Row>
            <Wizard.Main>
              {wizardSteps.map((step, stepIndex) => (
                <Wizard.Contents key={step.title} stepIndex={stepIndex} activeStepIndex={activeStepIndex}>
                  {wizardSteps[stepIndex].page}
                </Wizard.Contents>
              ))}
            </Wizard.Main>
          </Wizard.Row>
        </Wizard.Body>
        <Wizard.Footer>
          <Button
            bsStyle="default"
            className="btn-cancel"
            disabled={activeStepIndex === wizardSteps.length - 1}
            onClick={this.onCancel}
          >
            Cancel
          </Button>
          <Button
            bsStyle="default"
            disabled={activeStepIndex === 0 || activeStepIndex === wizardSteps.length - 1}
            onClick={this.onBack}
          >
            <Icon type="fa" name="angle-left" />Back
          </Button>
          {activeStepIndex < wizardSteps.length - 2 && (
            <Button bsStyle="primary" disabled={!stepOneValid} onClick={this.onNext}>
              Next<Icon type="fa" name="angle-right" />
            </Button>
          )}
          {activeStepIndex === wizardSteps.length - 2 && (
            <Button bsStyle="primary" disabled={!stepTwoValid} onClick={this.onSubmit}>
              Add
            </Button>
          )}
        </Wizard.Footer>
      </Wizard>
    );
  }
}

AccountWizard.propTypes = {
  show: PropTypes.bool.isRequired,
  edit: PropTypes.bool,
  error: PropTypes.bool,
  fulfilled: PropTypes.bool,
  stepOneValid: PropTypes.bool,
  stepTwoValid: PropTypes.bool,
  stepThreeValid: PropTypes.bool
};

AccountWizard.defaultProps = {
  edit: false,
  error: false,
  fulfilled: false,
  stepOneValid: false,
  stepTwoValid: false,
  stepThreeValid: false
};

const mapStateToProps = state => ({ ...state.accountWizard });

const ConnectedAccountWizard = connect(mapStateToProps)(AccountWizard);

export { ConnectedAccountWizard as default, ConnectedAccountWizard, AccountWizard };
