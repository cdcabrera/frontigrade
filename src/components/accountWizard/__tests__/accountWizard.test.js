import React from 'react';
import configureMockStore from 'redux-mock-store';
import { mount, shallow } from 'enzyme';
import { ConnectedAccountWizard, AccountWizard } from '../accountWizard';

describe('AccountView Component', () => {
  const generateEmptyStore = (obj = {}) => configureMockStore()(obj);

  it('should render a wizard', () => {
    const store = generateEmptyStore({ accountWizard: { show: true } });
    const component = shallow(<ConnectedAccountWizard />, { context: { store } });

    expect(component.dive()).toMatchSnapshot('connected');
  });

  it('should not display a wizard', () => {
    const props = {
      show: false
    };

    const component = mount(<AccountWizard {...props} />);
    expect(component.render()).toMatchSnapshot();
  });

  it('should have specific events defined', () => {
    const props = {
      show: false
    };
    const component = mount(<AccountWizard {...props} />);
    const componentInstance = component.instance();

    expect(componentInstance.onCancel).toBeDefined();
    expect(componentInstance.onNext).toBeDefined();
    expect(componentInstance.onBack).toBeDefined();
    expect(componentInstance.onSubmit).toBeDefined();
    expect(componentInstance.onStep).toBeDefined();
  });
});
