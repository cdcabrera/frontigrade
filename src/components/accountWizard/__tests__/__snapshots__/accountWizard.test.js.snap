// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountView Component should not display a wizard 1`] = `null`;

exports[`connected 1`] = `
<Wizard
  className=""
  dialogClassName=""
  onClose={[Function]}
  onExited={[Function]}
  show={true}
>
  <WizardHeader
    onClose={[Function]}
    title="Add Account"
  />
  <WizardBody
    className=""
  >
    <WizardSteps
      className=""
      steps={
        Array [
          <WizardStep
            activeStep={1}
            className=""
            label="1"
            onClick={[Function]}
            step={1}
            stepIndex={0}
            title="Policy"
          />,
          <WizardStep
            activeStep={1}
            className=""
            label="2"
            onClick={[Function]}
            step={2}
            stepIndex={1}
            title="Role"
          />,
          <WizardStep
            activeStep={1}
            className=""
            label="3"
            onClick={[Function]}
            step={3}
            stepIndex={2}
            title="ARN"
          />,
        ]
      }
    />
    <WizardRow
      className=""
    >
      <WizardMain
        className=""
      >
        <WizardContents
          activeStepIndex={0}
          activeSubStepIndex={null}
          className=""
          key="Policy"
          stepIndex={0}
          subStepIndex={null}
        >
          <Connect(AccountWizardStepPolicy)
            policySetupConfig="{
  \\"Version\\": \\"2012-10-17\\",
  \\"Statement\\": [
    {
      \\"Sid\\": \\"VisualEditor0\\",
      \\"Effect\\": \\"Allow\\",
      \\"Action\\": [
        \\"ec2:DescribeInstances\\",
        \\"ec2:DescribeImages\\",
        \\"ec2:DescribeSnapshots\\",
        \\"ec2:ModifySnapshotAttribute\\",
        \\"ec2:DescribeSnapshotAttribute\\"
      ],
      \\"Resource\\": \\"*\\"
    }
  ]
}"
          />
        </WizardContents>
        <WizardContents
          activeStepIndex={0}
          activeSubStepIndex={null}
          className=""
          key="Role"
          stepIndex={1}
          subStepIndex={null}
        >
          <AccountWizardStepTwo />
        </WizardContents>
        <WizardContents
          activeStepIndex={0}
          activeSubStepIndex={null}
          className=""
          key="ARN"
          stepIndex={2}
          subStepIndex={null}
        >
          <AccountWizardStepThree />
        </WizardContents>
      </WizardMain>
    </WizardRow>
  </WizardBody>
  <WizardFooter
    className=""
  >
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="default"
      className="btn-cancel"
      disabled={false}
      onClick={[Function]}
    >
      Cancel
    </Button>
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="default"
      disabled={true}
      onClick={[Function]}
    >
      <Icon
        name="angle-left"
        type="fa"
      />
      Back
    </Button>
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="primary"
      disabled={true}
      onClick={[Function]}
    >
      Next
      <Icon
        name="angle-right"
        type="fa"
      />
    </Button>
  </WizardFooter>
</Wizard>
`;
